src/rom_utils.cpp

```diff
@@ -137,34 +223,37 @@ bool load_rom(emu_shared_context_t *ctx, const char *filename) {
         std::cerr << "Could not read all of ROM image." << std::endl;
         free(full_image);
         stream.close();
-        return false;
+        return ROM_FAIL_READ;
     }
 
     stream.close();
 
-    //assume file is legitimate ROM
+    uint8_t load_rom_rc = validate_rom(full_image, rom_size);
+
+    if (load_rom_rc != ROM_OK) {
+        free(full_image);
+        return load_rom_rc;
```

Comment:
The rom still gets freed here even on the checksum error, which will crash since we still allow it to run.

(You can test with the roms at https://github.com/retrio/gb-test-roms/tree/master/cpu_instrs , which have broken checksums)